File: day-17/part-2/skasch.py
import bisect
import re
from typing import List, Optional, Set, Tuple

from tool.runners.python import SubmissionPy

REGEX = re.compile(r"target area: x=([-0-9]+)..([-0-9]+), y=([-0-9]+)..([-0-9]+)")


def find_vx(x1: int, x2: int, x_targets: List[int]) -> int:
    if x1 <= 0 <= x2:
        return 0
    if x2 < 0:
        m = -1
        x1, x2 = -x2, -x1
    else:
        m = 1
    idx1 = bisect.bisect_left(x_targets, x1)
    idx2 = bisect.bisect_left(x_targets, x2)
    if x_targets[idx1] == x1:
        return m * idx1
    if x_targets[idx2] == x2:
        return m * idx1
    if idx1 == idx2:
        raise ValueError
    return m * idx1


def find_vy(y1: int, y2: int) -> Tuple[int, bool]:
    vy1 = y1 if y1 >= 0 else -y1 - 1
    vy2 = y2 if y2 >= 0 else -y2 - 1
    if vy1 > vy2:
        return vy1, y1 >= 0
    return vy2, y2 >= 0


def reach_range_vx(
    ax1: int, ax2: int, vx: int, x_targets: List[int]
) -> Tuple[Optional[int], Optional[int]]:
    x_final = vx * (vx + 1) // 2
    if x_final < ax1 or vx > ax2:
        return None, None
    min_t = vx - bisect.bisect_right(x_targets, x_final - ax1) + 1
    if x_final <= ax2:
        return min_t, None
    max_t = vx - bisect.bisect_left(x_targets, x_final - ax2)
    if max_t < min_t:
        return None, None
    return min_t, max_t


def reach_range_y_in(y1: int, y2: int, steps: int) -> Set[int]:
    max_vy = (y2 + steps * (steps - 1) // 2) // steps
    min_vy = (y1 + steps * (steps - 1) // 2 + steps - 1) // steps
    return {vy for vy in range(min_vy, max_vy + 1)}


def parse(s: str) -> Tuple[int, int, int, int]:
    m = REGEX.match(s.strip())
    if m is not None:
        return (int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)))
    raise ValueError


class SkaschSubmission(SubmissionPy):
    def run(self, s: str) -> int:
        """
        :param s: input in string format
        :return: solution flag
        """
        # Your code goes here
        x1, x2, y1, y2 = parse(s)
        assert x1 * x2 >= 0
        if x2 <= 0:
            x1, x2 = -x2, -x1
        x_targets = [n * (n + 1) // 2 for n in range(x2 + 1)]
        res = 0
        min_vx = find_vx(x1, x2, x_targets)
        for vx in range(min_vx, x2 + 1):
            min_steps, max_steps = reach_range_vx(x1, x2, vx, x_targets)
            if min_steps is None:
                continue
            if max_steps is None:
                max_vy, _ = find_vy(y1, y2)
                max_steps = max_vy * 2 + 2
            vys = set()
            for steps in range(min_steps, max_steps + 1):
                vys |= reach_range_y_in(y1, y2, steps)
            res += len(vys)
        return res


def test_skasch() -> None:
    """
    Run `python -m pytest ./day-17/part-2/skasch.py` to test the submission.
    """
    assert (
        SkaschSubmission().run(
            """
target area: x=20..30, y=-10..-5
""".strip()
        )
        == 112
    )

File: day-17/part-2/skasch.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"sort"
	"strconv"
	"time"
)

var REGEX = regexp.MustCompile("target area: x=([-0-9]+)..([-0-9]+), y=([-0-9]+)..([-0-9]+)")

func build_x_targets(mx int) []int {
	x_targets := make([]int, mx+1)
	for n := 0; n <= mx; n++ {
		x_targets[n] = n * (n + 1) / 2
	}
	return x_targets
}

func find_vx(x1, x2 int, x_targets []int) int {
	if x1 <= 0 && 0 <= x2 {
		return 0
	}
	var m int
	if x2 < 0 {
		m = -1
		x1, x2 = -x2, x1
	} else {
		m = 1
	}
	idx1 := sort.Search(len(x_targets), func(i int) bool { return x_targets[i] >= x1 })
	idx2 := sort.Search(len(x_targets), func(i int) bool { return x_targets[i] >= x2 })
	if x_targets[idx1] == x1 {
		return m * idx1
	}
	if x_targets[idx2] == x2 {
		return m * idx1
	}
	if idx1 == idx2 {
		panic("Impossible to handle")
	}
	return m * idx1
}

func find_vy(y1, y2 int) int {
	var vy1, vy2 int
	if y1 >= 0 {
		vy1 = y1
	} else {
		vy1 = -y1 - 1
	}
	if y2 >= 0 {
		vy2 = y2
	} else {
		vy2 = -y2 - 1
	}
	if vy1 > vy2 {
		return vy1
	} else {
		return vy2
	}
}

func reach_range_vx(ax1, ax2, vx int, x_targets []int) (int, int) {
	x_final := vx * (vx + 1) / 2
	if x_final < ax1 || vx > ax2 {
		return -1, -1
	}
	min_t := vx + 1 - sort.Search(
		len(x_targets),
		func(i int) bool { return x_targets[i] > x_final-ax1 },
	)
	if x_final <= ax2 {
		return min_t, -1
	}
	max_t := vx - sort.Search(
		len(x_targets),
		func(i int) bool { return x_targets[i] >= x_final-ax2 },
	)
	if max_t < min_t {
		return -1, -1
	}
	return min_t, max_t
}

func floordiv(num, den int) int {
	rem := num % den
	if rem >= 0 {
		return num / den
	} else {
		return num/den - 1
	}
}

func reach_range_y_in(y1, y2, steps int) (int, int) {
	min_vy := floordiv(y1+steps*(steps+1)/2-1, steps)
	max_vy := floordiv(y2+steps*(steps-1)/2, steps)
	return min_vy, max_vy
}

func parse(s string) (int, int, int, int) {
	submatch := REGEX.FindStringSubmatch(s)
	x1, err := strconv.Atoi(submatch[1])
	if err != nil {
		panic(err)
	}
	x2, err := strconv.Atoi(submatch[2])
	if err != nil {
		panic(err)
	}
	y1, err := strconv.Atoi(submatch[3])
	if err != nil {
		panic(err)
	}
	y2, err := strconv.Atoi(submatch[4])
	if err != nil {
		panic(err)
	}
	return x1, x2, y1, y2
}

func run(s string) int {
	// Your code goes here
	x1, x2, y1, y2 := parse(s)
	if x1*x2 < 0 {
		panic("Failed to handle")
	}
	if x2 <= 0 {
		x1, x2 = -x2, -x1
	}
	x_targets := build_x_targets(x2)
	res := 0
	min_vx := find_vx(x1, x2, x_targets)
	for vx := min_vx; vx <= x2; vx++ {
		min_steps, max_steps := reach_range_vx(
			x1, x2, vx, x_targets,
		)
		if min_steps == -1 {
			continue
		}
		if max_steps == -1 {
			max_vy := find_vy(y1, y2)
			max_steps = max_vy*2 + 2
		}
		vys := make(map[int]interface{})
		for steps := min_steps; steps <= max_steps; steps++ {
			min_vy, max_vy := reach_range_y_in(y1, y2, steps)
			for nvy := min_vy; nvy <= max_vy; nvy++ {
				if _, ok := vys[nvy]; !ok {
					vys[nvy] = struct{}{}
					res += 1
				}
			}
		}
	}
	return res
}

func main() {
	// Uncomment this line to disable garbage collection
	// debug.SetGCPercent(-1)

	var input []byte
	var err error
	if len(os.Args) > 1 {
		// Read input from file for local debugging
		input, err = ioutil.ReadFile(os.Args[1])
		if err != nil {
			panic(err)
		}
		// Remove extra newline
		input = input[:len(input)-1]
	} else {
		// Read input from stdin
		input, err = ioutil.ReadAll(os.Stdin)
		if err != nil {
			panic(err)
		}
	}

	// Start resolution
	start := time.Now()
	result := run(string(input))

	// Print result
	fmt.Printf("_duration:%f\n", time.Since(start).Seconds()*1000)
	fmt.Println(result)
}
